### ISSUE TITLE
Differential input files

### ISSUE THREAD

**Issue by joshuahansel** [idaholab/moose#10676]
## Rationale
Minimize duplicate content between input files.

## Description
Command-line modification of input files is very handy; however, it has limitations. If I want to add or remove entire blocks/sub-blocks from an input file, this is not possible (or if I'm mistaken, it certainly wouldn't be a very pretty command line). If I want to change an object type in a block (having different parameters), this is not possible. As a result, I must duplicate input files. I'm proposing a syntax for applying modifications to an existing input file, for example:

```
# base_input_file.i

[SomeObjects]
  [./obj1]
    type = ClassA
    paramA = val1
    paramB = val2
  [../]
  [./obj2]
    type = ClassB
    paramC = val3
  [../]
[]

# (a lot of additional content here)
```
```
# modified_input_file.i

[BaseInputFile]
  relative_path = ./base_input_file.i
[]

[SomeObjects]
  [./obj1]
    # modify a parameter in the base input file
    paramB = val4
  [../]
  # remove a sub-block
  -[./obj2]
  # add an entirely new sub-block
  [./obj3]
    type = ClassC
    paramD = val5
  [../]
[]

# no need to copy all of the additional content
```

I do not know much about the parser, so I really don't know the feasibility/possibility of implementation, but I at least wanted to declare the desire for this capability.

If the usefulness of this capability still is not clear, here are just a few examples from RELAP-7:

- Change a fluid properties object
- Disconnect a component, connect a different one
- Change solution method, entailing a different preconditioner, time-stepper, etc.
- Change a stabilization object

There are so many instances in testing where I just want to test some kind of object in the context of an arbitrary physics problem, and I am forced to copy some input file that has physics that are not even relevant to the test (How many simple diffusion input files are in existence?). Every time I do this, it's one more file in which I'll need to maintain that "additional content".

In addition, even if command-line parameters are able to address a particular set of modifications, it is nice to actually have a physical file showing those changes (like if I want to re-run these modifications later). I extensively use the `cli_args` parameter for tester objects in `tests` files in the test suite, but if I want to run a particular test not represented by the base input file, then I need to be sure to replicate the command-line arguments each time (a small complaint, but frequent).

Anyway, I know this suggestion might be radical, but I wanted to get it out there in case there is shared interest and time/possibility to do this.

## Impact
Additional input functionality (no breakage of current functionality).

**Comment by dschwen** [comment]
It would not seem radical to allow for multiple `-i` options where later inputs override parameters from former inputs. This already works in a single file at the block level. Seems natural to me.

**Comment by joshuahansel** [comment]
Yeah you're right, I think that's a better way to do it than my suggestion of the `BaseInputFile` block. However, I think to achieve my goals above, we'd still need to develop the capability of _deleting_ blocks/parameters, not just _overwriting_  or _adding_ them.

**Comment by permcody** [comment]
Implementing an "include" mechanism has been discussed before and I'm open to the idea.

However, you can go further with today's implementation. Are you aware of the "active/inactive" parameter? You can put either one of those parameters in any block to either turn on/off whole blocks. Think of it as black/white lists of active blocks. We do this in a number of tests already.

**Comment by joshuahansel** [comment]
That's true, I had forgotten "active/inactive" existed. Yes, I guess as far as the "removal" capability, I think this would do the trick. The only disadvantage I can think of is that if I wanted to modify the type of an object block, I'd need to create a new block, with a new name (is this correct?). If so, I would need to make sure to manually modify any references to the original name. It wouldn't be ideal, but I could live with it.

**Comment by permcody** [comment]
So you don't have to change the name of the block, which would invalidate your references. You can change the type of a block which is super handy. You do have to be careful about unused parameters though when you start doing this, but it does work.

**Comment by joshuahansel** [comment]
I want to revive this discussion.

To summarize here, we currently have the ability to do the following:
* add a parameter
* modify an existing parameter's value
* effectively "delete" a block/sub-block by setting "inactive" to "true"

We do not have ability to do the following:
* remove a parameter

Is this right?

I'd really like to have the ability to remove a parameter. As Cody mentioned, if you change the type of a block, you may have unused parameter warnings, which we often promote to errors in testing. Also, in RELAP-7 at least, I sometimes just need to remove a parameter even though I'm not changing the type of the containing block. For instance, we sometimes disallow the user even specifying certain parameters in certain scenarios, so it's not enough to just set the parameter to its default parameter explicitly.

Could we have a command-line syntax for removing parameters, like maybe something like
```
myapp -i myfile.i -Executioner/solve_type
```
i.e., having a `-` in front of the parameter name?

**Comment by joshuahansel** [comment]
Regarding my original suggestion of the "differential input file", I like Daniel's suggestion, so long as we also allow testers to take a list of input files instead of a single file.

**Comment by joshuahansel** [comment]
@andrsd I'm having to completely duplicate some test problems input files because I'm switching spatial discretization, so I'm recommending implementation of some parameter-removal capability.

**Comment by permcody** [comment]
Let's bring @rwcarlsen in on this. I think we can support deleting parameters. I'm not sure if dash will work the best though. Many of those parameters are processed specially because that's the standard syntax for a flag. I'm sure we can come up with something though. I like dash, we'll just have to look at the implications.

**Comment by rwcarlsen** [comment]
I'd like to step back a bit and get a better feel for the high-level user experience issue(s) we are trying to solve before we get stuck discussing syntax and implementation details.  Can @joshuahansel provide a few *concrete* user stories of things he was trying to accomplish and how he ended up doing it with current MOOSE?

**Comment by rwcarlsen** [comment]
On a somewhat side note - I think we should be careful to not compromise provenance and reproducibility.  Fragmentation of content that we build inputs from has downsides (in addition to benefits).  Also, the concern of constantly copying irrelevant input from file to file just so you can test unrelated feature X suggests maybe we should put more effort into unit testing rather than trying to shoehorn too much into our regression-style test harness tests.

**Comment by joshuahansel** [comment]
Here's an example:
```
# Test 1: CG spatial discretization

# (a lot of problem setup)

[Components]
  [./pipeA]
    type = Pipe

    spatial_discretization = CG

    # (other pipe parameters)
  [../]

  # (other components)
[]
```
```
# Test 2: RDG spatial discretization

# (a lot of problem setup)

[Components]
  [./pipeA]
    type = Pipe

    spatial_discretization = RDG
    rdg_slope_reconstruction_type = MINMOD

    # (other pipe parameters)
  [../]

  # (other components)
[]
```
I'd like both tests to use the same input file, modified with CLI arguments. RELAP-7 throws an error if `rdg_slope_reconstruction_type` is provided by the user when `spatial_discretization = CG`. What do I do?

You might argue that the design of having parameters being illegal is some scenarios is flawed, that maybe I should have something `PipeCG` and `PipeRDG` to prevent this; however, this would not only be undesirable to a RELAP-7 user, but it wouldn't solve the issue: I'd still need to delete the `rdg_slope_reconstruction_type` parameter to avoid unused parameter errors.

To answer your question about how I solved this in current MOOSE, I didn't - I had to have 2 different, nearly identical input files.

**Comment by joshuahansel** [comment]
@rwcarlsen I agree with your argument regarding provenance (a word I had to look up by the way) and reproducibility. I don't know how much of an issue that would actually be, but I could see it as a possibility.

> Also, the concern of constantly copying irrelevant input from file to file just so you can test unrelated feature X suggests maybe we should put more effort into unit testing rather than trying to shoehorn too much into our regression-style test harness tests.

As much as I love unit tests, I don't see unit tests being an option here; in my example, I'm verifying the transient behavior of each of the spatial discretizations. Also, it is just not enough to test these things in a unit test setting. I need to make sure that all of objects are setup correctly in RELAP-7, not just they function correctly, given the correct setup.
